builtin/logical/transit/path_certificates.go

```diff
@@ -0,0 +1,345 @@
+	csrTemplate, err := parseCsr(pemCsrTemplate)
+	if err != nil {
+		return logical.ErrorResponse(err.Error()), logical.ErrInvalidRequest
+	}
+
+	pemCsr, err := p.CreateCsr(signingKeyVersion, csrTemplate)
+	if err != nil {
+		return nil, err
+	}
+
+	resp := &logical.Response{
+		Data: map[string]interface{}{
+			"name": p.Name,
+			"type": p.Type.String(),
+			"csr":  string(pemCsr),
+		},
+	}
+
+	return resp, nil
+}
+
+func (b *backend) pathImportCertChainWrite(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	name := d.Get("name").(string)
+
+	p, _, err := b.GetPolicy(ctx, keysutil.PolicyRequest{
+		Storage: req.Storage,
+		Name:    name,
+	}, b.GetRandomReader())
+	if err != nil {
+		return nil, err
+	}
+	if p == nil {
+		return logical.ErrorResponse(fmt.Sprintf("key with provided name '%s' not found", name)), logical.ErrInvalidRequest
+	}
+	if !b.System().CachingDisabled() {
+		p.Lock(true) // NOTE: Lock as we are might write to the policy
+	}
+	defer p.Unlock()
+
+	// Check if transit key supports signing
+	// NOTE: A key type that doesn't support signing cannot possible (?) have
+	// a certificate, so does it make sense to have this check?
+	if !p.Type.SigningSupported() {
+		return logical.ErrorResponse(fmt.Sprintf("key type %s does not support signing", p.Type)), logical.ErrInvalidRequest
+	}
+
+	// Check if key can be derived
+	if p.Derived {
+		return logical.ErrorResponse("operation not supported on keys with derivation enabled"), logical.ErrInvalidRequest
+	}
+
+	// Transit key version
+	keyVersion := p.LatestVersion
+	if version, ok := d.GetOk("version"); ok {
+		keyVersion = version.(int)
+	}
+
+	// Get certificate chain
+	pemCertChain := d.Get("certificate_chain").(string)
+	certChain, err := parseCertificateChain(pemCertChain)
+	if err != nil {
+		return logical.ErrorResponse(err.Error()), logical.ErrInvalidRequest
+	}
+
+	err = validateLeafCertPosition(certChain)
+	if err != nil {
+		return logical.ErrorResponse(err.Error()), logical.ErrInvalidRequest
+	}
+
+	leafCertPublicKeyAlgorithm := certChain[0].PublicKeyAlgorithm
+
+	// Check if end-entity public key algorithm matches transit key
+	err = validateKeyTypeMatch(p, leafCertPublicKeyAlgorithm)
+	if err != nil {
+		return logical.ErrorResponse(err.Error()), logical.ErrInvalidRequest
+	}
+
+	// Validate if leaf cert key matches with transit key
+	valid, err := p.ValidateLeafCertKeyMatch(keyVersion, leafCertPublicKeyAlgorithm, certChain[0].PublicKey)
+	if err != nil {
+		return nil, fmt.Errorf("could not validate key match between leaf certificate key and key version in transit: %w", err)
+	}
+	if !valid {
+		return logical.ErrorResponse("leaf certificate public key does match the key version selected"), logical.ErrInvalidRequest
+	}
+
+	err = p.PersistCertificateChain(ctx, keyVersion, certChain, req.Storage)
+	if err != nil {
+		return nil, fmt.Errorf("failed to persist certificate chain: %w", err)
+	}
+
+	return nil, nil
```

Comment:
Do you want to show the persisted CA chain here in the response? We tend to like write responses to include the written data these days in cryptosec projects at least. :-)

Reply:
Yes, I like that idea. For now, I set up the certificate chain, "certificate-chain", to be returned as a `string`. I would prefer returning a slice of strings but ended up going with `string` to follow the same behavior as the export endpoint. Thoughts?

Reply:
Nah, I don't think storing the formatted certificate saves much compute tbh. Its minimal overhead and we have a lot more overhead in other places IMO.